src/firmware/app/control/trajectory_planner.h

```diff
@@ -0,0 +1,271 @@
+    float time;
+} TrajectoryElement_t;
+
+typedef struct Trajectory
+{
+    TrajectoryElement_t* trajectory_elements;
+    unsigned int num_elements;
+} Trajectory_t;
+
+typedef enum TrajectoryPlannerGenerationStatus
+{
+    OK                        = 0,
+    FINAL_VELOCITY_TOO_HIGH   = 1,
+    INITIAL_VELOCITY_TOO_HIGH = 2,
+} TrajectoryPlannerGenerationStatus;
+
+typedef struct FirmwareRobotPathParameters
+{
+    // The 2D polynomial representation of the path to be followed
+    Polynomial2dOrder3_t path;
+    // The path parameterization value indicating the beginning of the
+    // considered path
+    float t_start;
+    // The path parameterization value indicating the end of the considered
+    // path
+    float t_end;
+    // The number of segments to discretize the trajectory into.
+    //  Must be greater than 2 segments.
+    //  THE NUMBER OF SEGMENTS MUST BE UNDER
+    //  TRAJECTORY_PLANNER_MAX_NUMBER_SEGMENTS
+    unsigned int num_segments;
+    // The maximum acceleration allowed at any
+    //  point along the trajectory. This factor limits the maximum delta-velocity
+    // and
+    //  also the max speed around curves due to centripetal acceleration [m/s^2]
+    float max_allowable_acceleration;
+    // The maximum speed allowable at any point along the trajectory```
+    float max_allowable_speed;
+    // The initial speed at the start of the trajectory [m/s]
+    float initial_speed;
+    // The final speed at the end of the trajectory [m/s]
+    float final_speed;
+} FirmwareRobotPathParameters_t;
+
+/**
+ * Returns a planned trajectory with the list of guarantees based on the assumptions below
+ *
+ *  Key assumptions & guarantees of this planner are:
+ *  - Trajectories are time-optimal assuming INFINITE JERK capability of the robot
+ *
+ *  - No speed on the trajectory is larger than the 'max_allowable_speed' parameter
+ *
+ *  - No acceleration value between points on the trajectory can be larger than the
+ * 'max_allowable_acceleration' input parameter
+ *
+ *  - Assuming the grip-limit of the robot IS THE SAME AS THE MAX ACCELERATION then at no
+ * point on the path can the sum of centripetal and acceleration force be greater than
+ * 'max_allowable_acceleration'
+ *
+ *  Trajectory generation is done by assuming constant acceleration capability.
+ *      - The generator will assume maximum acceleration for the robot between each
+ * segment on the path. If using max acceleration breaches either the speed limit of the
+ * robot, or the centripetal acceleration limit defined by the curvature and robot speed -
+ * the planner will assume the max acceleration that will remain bellow this limit.
+ *
+ *     - To ensure the robot is capable of deccelerating to react to any sudden changes in
+ * the path, the trajectory is checked for backwards continuity
+ *
+ * NOTE: The following function is based on an algorithm described in
+ * http://www2.informatik.uni-freiburg.de/~lau/students/Sprunk2008.pdf
```

Comment:
Can you point to a specific algorithm name or page number, since that paper is very long?

Reply:
Added page # :+1: